File: Add Code Here/C++/mergesort.cpp
#include <iostream>
using namespace std;

void merge(int *Arr, int st, int mid, int en){
    int B[(en-st+1)];
    int length = en-st+1;
    int cst=st, mst=mid+1, count=0;
    while(count<length){
        if(cst<=mid){
            if(mst<=en){
                if(Arr[cst]<Arr[mst]){
                    B[count++] = Arr[cst++];
                }else{
                    B[count++] = Arr[mst++];
                }
            }else{
                B[count++] = Arr[cst++];
            }
        }else{
            if(mst<=en){
                B[count++]= Arr[mst++];
            }
        }
    }
    for(int i=0;i<length;i++){
        Arr[i+st] = B[i];
    }
}

void mergesort(int* Arr, int st, int en){
    if(en>st){
        int mid = (st+en)/2;
        mergesort(Arr, st, mid);
        mergesort(Arr, mid+1, en);
        merge(Arr, st, mid, en);
    }
}

void sort(int* Arr, int length){
    mergesort(Arr, 0, length-1);
}

int main(){
    int n;
    cout<<"Enter the length of array\n";
    cin>>n;
    int Arr[n];
    for(int i = 0;i<n;i++){
        cout<<"Enter an array element\n";
        cin>>Arr[i];
    }
    //sort
    sort(Arr, n);

    //Print after sorting
    cout<<"Sorted array\n";
    for(int j=0;j<n;j++){
        cout<<Arr[j]<<endl;
    }
    return(0);
}

File: Add Code Here/Java/merge_sort.java
package Sorting;

import java.util.Scanner;

public class merge_sort {
    static void mergeSort(int arr[], int start, int end)
    {
        int mid;
        if(start<end){
            mid=(start+end)/2;
            mergeSort(arr,start,mid);
            mergeSort(arr,mid+1,end);
            merge(arr,start,mid,end);
        }

    }
    static void merge(int arr[],int start,int mid, int end)
    {
        int [] b=new int[100];
        int i=start,j=mid+1,k=start;
        while(i<=mid && j<=end){
            if(arr[i]<=arr[j]){
                b[k++]=arr[i++];
            }
            else
                b[k++]=arr[j++];
        }
        if(i>mid){
            while(j<=end){
                b[k++]=arr[j++];
            }
        }
        else{
            while(i<=mid)
                b[k++]=arr[i++];
        }
        for(i=start;i<=end;i++){
            arr[i]=b[i];
        }
    }


    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the size of array: ");
        int n = input.nextInt();
        int [] arr = new int[n];
        int i;
        System.out.print("Enter "+ n + " element of array: ");
        for(i=0;i<n;i++)
            arr[i]= input.nextInt();
        mergeSort(arr,0,n-1);
        System.out.print("Sorted elements are: ");
        for(i=0;i<n;i++)
            System.out.print(arr[i]+ " ");
    }
}
